engine/access/rest/websockets/controller_test.go

```diff
@@ -0,0 +1,252 @@
+package websockets
+
+import (
+	"context"
+	"encoding/json"
+	"testing"
+
+	"github.com/google/uuid"
+	"github.com/gorilla/websocket"
+	"github.com/rs/zerolog"
+	"github.com/stretchr/testify/mock"
+	"github.com/stretchr/testify/require"
+	"github.com/stretchr/testify/suite"
+
+	dpmock "github.com/onflow/flow-go/engine/access/rest/websockets/data_provider/mock"
+	connmock "github.com/onflow/flow-go/engine/access/rest/websockets/mock"
+	"github.com/onflow/flow-go/engine/access/rest/websockets/models"
+	"github.com/onflow/flow-go/engine/access/state_stream/backend"
+	streammock "github.com/onflow/flow-go/engine/access/state_stream/mock"
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/utils/unittest"
+)
+
+type WsControllerSuite struct {
+	suite.Suite
+
+	logger       zerolog.Logger
+	wsConfig     Config
+	streamApi    *streammock.API
+	streamConfig backend.Config
+}
+
+func (s *WsControllerSuite) SetupTest() {
+	s.logger = unittest.Logger()
+	s.wsConfig = NewDefaultWebsocketConfig()
+	s.streamApi = streammock.NewAPI(s.T())
+	s.streamConfig = backend.Config{}
+}
+
+func TestWsControllerSuite(t *testing.T) {
+	suite.Run(t, new(WsControllerSuite))
+}
+
+// TestSubscribeRequest tests the subscribe to topic flow.
+// We emulate a request message from a client, and a response message from a controller.
+func (s *WsControllerSuite) TestSubscribeRequest() {
+	s.T().Run("Happy path", func(t *testing.T) {
+		conn, dataProviderFactory, dataProvider := newControllerMocks(t)
+		controller := NewWebSocketController(s.logger, s.wsConfig, dataProviderFactory, conn)
+
+		dataProvider.
+			On("Run", mock.Anything).
+			Run(func(args mock.Arguments) {}).
+			Once()
+
+		requestMessage := models.SubscribeMessageRequest{
+			BaseMessageRequest: models.BaseMessageRequest{Action: "subscribe"},
+			Topic:              "blocks",
+			Arguments:          nil,
+		}
+
+		// Simulate receiving the subscription request from the client
+		conn.
+			On("ReadJSON", mock.Anything).
+			Run(func(args mock.Arguments) {
+				reqMsg, ok := args.Get(0).(*json.RawMessage)
+				require.True(t, ok)
+				msg, err := json.Marshal(requestMessage)
+				require.NoError(t, err)
+				*reqMsg = msg
```

Comment:
what does this do?

Reply:
This is a signature of `ReadJSON(v interface{})`. This code sets the `requestMessage` to `v`